source/vibe/http/server.d

```diff
@@ -730,44 +897,133 @@ final class HTTPServerRequest : HTTPRequest {
 */
 final class HTTPServerResponse : HTTPResponse {
 	private {
-		Stream m_conn;
-		ConnectionStream m_rawConnection;
-		OutputStream m_bodyWriter;
+		union Connection {
+			ConnectionStack stack;
+			ConnectionStream test;
+		}
+
+		struct ConnectionStack {
+			TCPConnection tcp;
+			TLSStream tls;
+			HTTP2Stream http2;
+		}
+
+		Connection m_conn;
+		CompressionStream m_compressionStream;
+		BodyStream m_bodyStream;
+
 		Allocator m_requestAlloc;
-		FreeListRef!ChunkedOutputStream m_chunkedBodyWriter;
-		FreeListRef!CountingOutputStream m_countingWriter;
-		FreeListRef!GzipOutputStream m_gzipOutputStream;
-		FreeListRef!DeflateOutputStream m_deflateOutputStream;
 		HTTPServerSettings m_settings;
 		Session m_session;
-		bool m_headerWritten = false;
-		bool m_isHeadResponse = false;
-		bool m_ssl;
+
+		bool m_isTest;
+		bool m_isGzip;
+		bool m_isChunked;
+		bool m_headerWritten;
+		bool m_isHeadResponse;
+		bool m_outputStream;
 		SysTime m_timeFinalized;
 	}
 
-	this(Stream conn, ConnectionStream raw_connection, HTTPServerSettings settings, Allocator req_alloc)
+	union CompressionStream {
+		GzipOutputStream gzip;
+		DeflateOutputStream deflate;
+	}
+
+	union BodyStream {
+		/// Counts content-length to validate the value specified in the headers
+		CountingOutputStream counting;
+		/// Sends chunks of data after every flush
+		ChunkedOutputStream chunked;
+		/// For HEAD Requests
+		NullOutputStream none;
+	}
+
+	/// Identifies the stream most suitable for writing through
+	private {
+		@property OutputStream outputStream() {
+			if (!m_outputStream) return null;
+			if (!hasCompression) {
+				if (isHeadResponse) return cast(OutputStream) m_bodyStream.none;
+				else if (m_isChunked) return cast(OutputStream) m_bodyStream.chunked;
+				else return cast(OutputStream) m_bodyStream.counting;
+			}
+			else if (m_isGzip) return cast(OutputStream) m_compressionStream.gzip;
+			else return cast(OutputStream) m_compressionStream.deflate;
+		}
+
+		@property ConnectionStream topStream() { return m_isTest ? cast(ConnectionStream) m_conn.test : (m_conn.stack.http2 ? cast(ConnectionStream) m_conn.stack.http2 : ( m_conn.stack.tls ? cast(ConnectionStream) m_conn.stack.tls : cast(ConnectionStream) m_conn.stack.tcp)); }
```

Comment:
That's some serious abuse of ternary operator here :D
Can you make this nested `if` please ?